WordPress/openverse-catalog
# Update Flickr to use new time delineated ingester class

## Fixes




Fixes #26 by @obulat 

## Description



This PR is largely a refactor and has very little new logic. The goal is to update the Flickr DAG to dynamically break its ingestion day down into time intervals depending on the distribution of data throughout the day. To do so, I've pulled existing logic out of the Finnish Museums DAG into a reusable class to be used by both.

Summary:
* Creates a subclass of the `ProviderDataIngester`, `TimeDelineatedProviderDataIngester`, which adds in all the logic for generating timestamp intervals, iterating over them, and detecting some common bugs
* Refactors Finnish and Flickr to use this base class

To use the new class all you have to do is define a few variables and implement an abstract method for getting the record count for a particular interval. When `ingest_records` is called, timestamp pairs will be dynamically generated according to the configured variables, and iterated over automatically. The bounds of the current time interval will `start_ts` and `end_ts` will be available as kwargs in `get_query_params`.

The behavior of Finnish Museums should be totally unchanged. Flickr now gets for free the dynamic timeslice generation that Finnish was already working with. It will attempt to generate intervals with less than 4,000 records per slice, as the Flickr documentation reports that maximum 4k unique records are returned for any given query.

### Notes about how the timeslice intervals were chosen

When experimenting locally, I found that the benefit of decreasing the interval size rapidly diminishes for Flickr (ie, after a certain point, you start getting _far more_ duplicates when the intervals decrease in size). I suspect that the API isn't actually meant to be used for very small intervals, so the range is 'fuzzy' and causes a lot of overlap/duplicates when these small intervals are used.

For reference, I tried running ingestion using the date `2023-02-08` using (1) 1-minute time slices for the entire day, and (2) the dynamic interval generation used in this PR.

For the first, that resulted in **1440** intervals and took almost **1.5 hours** to pull data. We received **37,115** unique records and a whopping **709,199** duplicates.

For the latter, only **35** intervals were generated and it took just under **3 minutes** to pull (1/30 of the time). We got **34,768** records but only **18,295** duplicates.

So, for the cost of a relatively small difference in the number of records ingested we are able to run ingestion in a fraction of the time and deal with many, many fewer duplicates. It is also worth pointing out that neither approach guarantees that all unique results are collected, so we are making a concession either way. While we pursue other approaches to Flickr, I think we should go with the much faster approach, and perhaps tweak the reingestion schedule for Flickr to be more aggressive.


## Testing Instructions


Run tests locally.

Manually test the Flickr and Finnish PRs. For Flickr, you should be able to simply enable it locally and let it run for a few days -- it should be pretty quick! For Finnish you may need to let it run for awhile to make sure that it hits an ingestion day with a reasonable amount of data, as that DAG is quite sparse.


## Checklist



      `Update index.md`).
      a parent feature branch.
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin



<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
